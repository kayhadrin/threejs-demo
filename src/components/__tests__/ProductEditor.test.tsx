import { act, render, screen } from '@testing-library/react';
// @ts-expect-error: maybe related to https://github.com/testing-library/user-event/discussions/978#discussioncomment-3359165 ?
import userEvent from '@testing-library/user-event';
import ProductEditorOriginal from '../ProductEditor';
import './ProductEditorCanvasMock';
import '../3DModels/__tests__/Lazy3DModelsMock';

describe('ProductEditor component', () => {
  let ProductEditor: typeof ProductEditorOriginal;
  beforeEach(async () => {
    ProductEditor = (await import('../ProductEditor')).default;
  });

  describe('when the user selects a container material and container template', () => {
    // Warning: this test logs some console.errors but it doesn't actually fail the test expectations
    // Message looks like: `The current testing environment is not configured to support act(...)`
    it('renders a 3D model', async () => {
      const user = userEvent.setup();
      let container;
      await act(async () => {
        container = render(<ProductEditor />).container;
      });
      await act(async () => {
        await user.selectOptions(
          screen.getByRole('combobox', { name: /Container Material/i }),
          'container-material-1'
        );
      });
      await act(async () => {
        await user.click(screen.getByAltText(/Shampoo bottle/i));
      });
      expect(screen.getByTestId('lazy3DModel-ShampooBottle3DModel')).toBeVisible();
      expect(container).toMatchSnapshot();
    });
  });
});
