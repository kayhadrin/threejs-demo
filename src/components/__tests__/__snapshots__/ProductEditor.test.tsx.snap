// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductEditor component when the user selects a container material and container template renders a 3D model 1`] = `
<div>
  <div
    style=""
  >
    [[[ProductEditorCanvas
    <div
      data-testid="lazy3DModel-ShampooBottle3DModel"
    >
      [[[ShampooBottle3DModel
      props = {}
      ShampooBottle3DModel]]]
    </div>
    ProductEditorCanvas]]]
  </div>
  <div
    class="z-0 rounded bg-blue-100/50 p-4 sm:p-8 md:p-12"
  >
    <a
      class="text-blue-500 hover:underline"
      href="/"
    >
      Home
    </a>
    /
    <h1
      class="text-3xl font-bold tracking-tight sm:text-5xl"
    >
      Product Editor
    </h1>
    <section
      class="flex flex-col gap-4"
    >
      <label
        class="flex flex-col"
      >
        <span
          class="mb-1 font-medium"
        >
          Container Material
        </span>
        <select
          class="rounded border px-2 py-1"
          name="container-material-dropdown"
          style="background-color: rgba(255, 255, 255, 0.8);"
        >
          <option
            value=""
          >
            Select a material
          </option>
          <option
            value="container-material-1"
          >
            Plastic
          </option>
          <option
            value="container-material-2"
          >
            Glass
          </option>
        </select>
      </label>
      <label
        class="flex flex-col"
      >
        <span
          class="mb-1 font-medium"
        >
          Container Template
        </span>
        <div
          class="flex flex-wrap gap-4"
        >
          <button
            class="flex flex-col items-center rounded border p-2 transition ring-2 ring-blue-500"
            style="background: rgba(59, 130, 246, 0.1);"
            title="Select Shampoo Bottle"
            type="button"
          >
            <img
              alt="Shampoo Bottle"
              class="mb-2 h-24 w-24 object-contain"
              data-nimg="1"
              decoding="async"
              height="496"
              loading="lazy"
              src="/_next/image?url=%2F3DModels%2FshampooBottle%2Fscene.thumbnail.png&w=1080&q=75"
              srcset="/_next/image?url=%2F3DModels%2FshampooBottle%2Fscene.thumbnail.png&w=640&q=75 1x, /_next/image?url=%2F3DModels%2FshampooBottle%2Fscene.thumbnail.png&w=1080&q=75 2x"
              style="color: transparent;"
              width="468"
            />
          </button>
        </div>
      </label>
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="flex items-center align-middle"
        >
          <span
            class="inline-block max-h-lh max-w-sm overflow-hidden bg-white p-1 font-mono overflow-ellipsis whitespace-nowrap text-blue-500 select-all"
          >
            http://localhost/?containerMaterialID=container-material-1&containerTemplateID=container-template-1
          </span>
          <a
            class="ml-2 rounded border-2 border-blue-500 bg-blue-100 p-1 transition-all duration-300 ease-out hover:bg-blue-200 active:border-green-500 active:bg-green-100"
            href="http://localhost/?containerMaterialID=container-material-1&containerTemplateID=container-template-1"
          >
            Share
          </a>
        </div>
      </div>
    </section>
  </div>
</div>
`;
